cpp/src/io/text/bgzip_data_chunk_source.cu

```diff
@@ -0,0 +1,451 @@
+ public:
+  device_uvector_data_chunk(rmm::device_uvector<char>&& data) : _data(std::move(data)) {}
+
+  [[nodiscard]] char const* data() const override { return _data.data(); }
+  [[nodiscard]] std::size_t size() const override { return _data.size(); }
+  operator device_span<char const>() const override { return _data; }
+
+ private:
+  rmm::device_uvector<char> _data;
+};
+
+struct bgzip_nvcomp_transform_functor {
+  uint8_t const* compressed_ptr;
+  uint8_t* decompressed_ptr;
+
+  __device__ thrust::tuple<device_span<const uint8_t>, device_span<uint8_t>> operator()(
+    thrust::tuple<std::size_t, std::size_t, std::size_t, std::size_t> t)
+  {
+    auto const compressed_begin   = thrust::get<0>(t);
+    auto const compressed_end     = thrust::get<1>(t);
+    auto const decompressed_begin = thrust::get<2>(t);
+    auto const decompressed_end   = thrust::get<3>(t);
+    return thrust::make_tuple(device_span<const uint8_t>{compressed_ptr + compressed_begin,
+                                                         compressed_end - compressed_begin},
+                              device_span<uint8_t>{decompressed_ptr + decompressed_begin,
+                                                   decompressed_end - decompressed_begin});
+  }
+};
+
+class bgzip_data_chunk_reader : public data_chunk_reader {
+ private:
+  /*
+
+
+     Parsing code
+
+
+  */
+
+  template <typename IntType>
+  static bool is_little_endian()
+  {
+    IntType i = 0x0100;
+    std::array<char, sizeof(i)> bytes;
+    std::memcpy(&bytes[0], &i, sizeof(i));
+    return bytes[0] == 0;
+  }
+
+  template <typename IntType>
+  static IntType read_int(char* data)
+  {
+    IntType result{};
+    if (not is_little_endian<IntType>()) { std::reverse(data, data + sizeof(result)); }
+    std::memcpy(&result, &data[0], sizeof(result));
+    return result;
+  }
+
+  struct bgzip_header {
+    int block_size;
+    int extra_length;
+    [[nodiscard]] int data_size() const { return block_size - extra_length - 20; }
+  };
+
+  bgzip_header read_header()
+  {
+    std::array<char, 12> buffer{};
+    CUDF_EXPECTS(_stream->read(buffer.data(), sizeof(buffer)), "read failed");
+    uint8_t constexpr magic_expected_2 = 139;
+    std::array<char, 4> expected_header{{31, 0, 8, 4}};
+    std::memcpy(&expected_header[1], &magic_expected_2, sizeof(char));
+    CUDF_EXPECTS(std::equal(expected_header.begin(), expected_header.end(), buffer.begin()),
+                 "malformed BGZIP header");
+    auto extra_length = read_int<uint16_t>(&buffer[10]);
+    uint16_t extra_offset{};
+    // read all the extra subfields
+    while (extra_offset < extra_length) {
+      auto const remaining_size = extra_length - extra_offset;
+      CUDF_EXPECTS(remaining_size >= 4, "invalid extra field length");
+      CUDF_EXPECTS(_stream->read(buffer.data(), 4), "read failed");
+      extra_offset += 4;
+      auto subfield_size = read_int<uint16_t>(&buffer[2]);
+      if (buffer[0] == 66 && buffer[1] == 67) {
```

Comment:
A comment to explain these values would be nice here. Otherwise the reader has to look into http://samtools.github.io/hts-specs/SAMv1.pdf.

Reply:
I'll add a description of the fields